pytorch_lightning/loops/closure.py

```diff
@@ -0,0 +1,113 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Callable, Optional
+
+from torch import Tensor
+
+from pytorch_lightning.profiler import BaseProfiler
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+@dataclass
+class ClosureResult:
+    closure_loss: Optional[Tensor]
+    result_collection: Optional[ResultCollection]
+
+    @property
+    def loss(self) -> Optional[Tensor]:
+        if self.closure_loss is not None:
+            return self.closure_loss.detach().clone()
+
+
+class Closure(ABC):
+    """
+    Abstract base class for optimizer closures in Lightning.
+
+    Formally, a closure is binding variables from an external scope to a function that does a computation on these
+    variables without taking them explicitly as input. This has the benefit that a closure can be passed to an
+    object which later can call it like a function but without requiring to pass in any arguments.
+
+    This class provides a simple abstraction making the instance of this class callable like a function while capturing
+    the :class:`ClosureResult` and caching it.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._result = None
+
+    @property
+    def result(self) -> ClosureResult:
+        """The cached result from the last time the closure was called."""
+        return self._result
+
+    @abstractmethod
+    def closure(self, *args: Any, **kwargs: Any) -> ClosureResult:
+        """Implements the behavior of the closure once it is getting called."""
+        pass
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Optional[Tensor]:
+        self._result = self.closure(*args, **kwargs)
+        if self._result is not None:
+            return self._result.loss
+
+
+class LightningClosure(Closure):
```

Comment:
Naming (see above): DefaultLightningClosure

Reply:
How about Closure and ClosureBase? Recently @ananthsub raised concern that we name too many internal components prefixed with Lightning*